CFG Cleanup pass - Remove unreachable blocks.

dnovillo opened the pull request:
This change adds a new CFG cleanup pass to spirv-opt.  The intent of the pass is to act as a repository of a variety of common CFG cleanup actions (straightening, removal of unreachable code, etc).

It's my first change to SPIRV-Tools, so I'm sure it's full of inconsistencies and missing bits.  One of the things I've been thinking is that it may make sense to coalesce other existing CFG cleanup passes into this one in the hopes of reducing the number of CFG/IR traversals needed during cleanup.

Or, perhaps, it's preferable to keep these passes separate.  I understand that compile time is not an issue now, but I'm not sure if this might be challenging in the future (in my experience, this tends to be true at some point).

The code works and passes the trivial test I've added.  I need to add more, but I want a first review pass so I can learn preferred idioms and ways of structuring the code.

Thanks.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/SPIRV-Tools?pullRequest=810) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/SPIRV-Tools?pullRequest=810) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/SPIRV-Tools?pullRequest=810) it.</sub>

dneto0 commented:
I haven't looked at the code yet, but am curious about "straightening".
Something that has been requested is:
 - SPIR-V has a rule requiring that if block A dominates block B, then A appears in the module before B.
 - It would be nice if we had a pass that enforced:  If B post-dominates A, then B appears after A.
 - Furthermore, it would be nice if, when structured control flow is in effect, the continue construct for a loop appears after the loop construct for the loop. (See section 2.11 for those definitions)

greg-lunarg commented:
A few notes on the most recent comment on straightening by @dneto0:

- Another order "nicety" (I think) would be to have the merge block follow all blocks in the control-flow construct  and continue construct for which it is the merge block.

- There is code in spirv-opt that figures out an ordering that satisfies this requirement plus those from @dneto0. Look for ComputeStructuredOrder(). Note that there are several copies laying around. I think these can (and should) be consolidated into one at some point.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,88 @@
+  // Erase unreachable blocks.
+  for (auto ebi = func->begin(); ebi != func->end();)
+    if (reachable_blocks.find(&*ebi) == reachable_blocks.end()) {
+      ebi = ebi.Erase();
```
You're too used to LLVM.  Remember that in the current ir::* class hierarchy, instructions only have IDs for the things they use, and the DefUseManager is a separate layer over it.

So deleting an Instruction will not delete the references to that instruction.
Thinks you'll have to clean up:
 - OpName,  possibly OpMemberName (for struct types, so does not apply in this pass)
 - OpDecoration, OpDecorationGroup
 - OpPhi, e.g. you delete A, and so all values defined in A, but that block and its values might be used in an OpPhi in a live block.

Use DefUseManager to trace uses.  It has a convenience KillInst that converts an existing instruction to a Nop and kill the uses.  That'a a common idiom you'll see:  KillInst, then the Module's write-to-output function can be used to skip over all nops.  It works.  However, cleaning up blocks is another matter.   Check Greg's dead-branch-elim code for examples of how to clean up blocks.

And then we have to agree on whether this pass promises to keep up the DefUseManager up to date. By default it does not have to make that promise.  But our direction is to "save time" by agreeing to keep it up to date, and then have the pass manager track whether the DefUseManager is stale or not.

dneto0 commented on include/spirv-tools/optimizer.hpp:
```diff
@@ -393,6 +393,13 @@ Optimizer::PassToken CreateAggressiveDCEPass();
 
+// Creates a CFG cleanup pass.
+// This pass removes cruft from the control flow graph. It currently includes
+// the following functionality:
```
The pass only operates on functions reachable from some entry point.  Should say so.

dnovillo replied:
Done.

dneto0 commented on test/opt/cfg_cleanup_test.cpp:
```diff
@@ -0,0 +1,88 @@
+
+  SinglePassRunAndCheck<opt::CFGCleanupPass>(
+      declarations + body_before, declarations + body_after, true, true);
+}
```
Need cases for:
 - showing cleanup of OpName and OpDecoration, OpGroupDecorate
 - showing cleanup of OpPhis in live blocks that remain
 - keeping unreachable merge block
 - keeping unreachable continue target, and the loop backedge it dominates even if it's not in the same block
 - in these latter two cases: non-trivial defs and uses inside those unreachable blocks

dneto0 commented on tools/opt/opt.cpp:
```diff
@@ -112,9 +112,12 @@ NOTE: The optimizer is a work in progress.
-  -h, --help   
+  --cfg-cleanup
+               Cleanup the control flow graph. This will remove any unnecessary
+               code from the CFG like unreachable code.
```
.... "Peformed only on entry point call tree functions"

dneto0 commented:
I think this PR is still live but needs feedback addressed.

dneto0 commented:
Also, please rebase to avoid conflicts.

dnovillo commented:
On Wed, Sep 20, 2017 at 4:23 PM, David Neto <notifications@github.com>
wrote:

> I think this PR is still live but needs feedback addressed.
>

It is, but I've been traveling and distracted with shiny things.  Will
respond soon.

dnovillo commented:
On Fri, Sep 8, 2017 at 1:23 PM, David Neto <notifications@github.com> wrote:

> You also have to preserve any Merge block or ContinueTarget mentioned by
> an OpLoopMerge or OpSelectionMerge instruction, even if they are
> unreachable by control flow.
>
> BasicBlock::ForMergeAndContinueLabel is your friend here.
>

Thanks.  Done.


> Now, if the merge block is unreachable, then
>
>    - you can delete that merge blocks' instructions. (Reasonable to defer
>    to a
>
> ... future patch?


>
>    - the merge block's own CFG successors might not be reachable either.
>    Similar for the ContinueTarget, except that you must also preserve the
>    loop backedge, and that loop backedge might not be in the same block as the
>    ContinueTarget.
>
> But if they are not reachable, then they would be caught by the other
logic.  Perhaps I'm just not following you here.


>
>
> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137830811>
> :
>
> > +bool CFGCleanupPass::RemoveUnreachableBlocks(ir::Function* func) {
> +  bool modified = false;
> +
> +  // Mark reachable all blocks reachable from the function's entry block.
> +  std::unordered_set<ir::BasicBlock*> reachable_blocks;
> +  std::queue<ir::BasicBlock*> worklist;
> +  reachable_blocks.insert(func->entry().get());
> +  worklist.push(func->entry().get());
> +  while (!worklist.empty()) {
> +    ir::BasicBlock* block = worklist.front();
> +    worklist.pop();
> +    block->ForEachSuccessorLabel(
> +        [&reachable_blocks, &worklist, this](uint32_t label_id) {
> +          auto successor = label2block_[label_id];
> +          reachable_blocks.insert(successor);
> +          worklist.push(successor);
>
> I think you should only push it onto the worklist if it wasn't already in
> the reachable blocks set.
>
> Does this code terminate when it sees a loop?
>
It doesn't.  Fixed.

------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137831660>
> :
>
> > +  reachable_blocks.insert(func->entry().get());
> +  worklist.push(func->entry().get());
> +  while (!worklist.empty()) {
> +    ir::BasicBlock* block = worklist.front();
> +    worklist.pop();
> +    block->ForEachSuccessorLabel(
> +        [&reachable_blocks, &worklist, this](uint32_t label_id) {
> +          auto successor = label2block_[label_id];
> +          reachable_blocks.insert(successor);
> +          worklist.push(successor);
> +        });
> +  }
> +
> +  // Erase unreachable blocks.
> +  for (auto ebi = func->begin(); ebi != func->end();)
> +    if (reachable_blocks.find(&*ebi) == reachable_blocks.end()) {
>
> Some prefer using the count method instead of check against end iterator.
>
> if (reachable.count(&*ebi))) { ++ebi ; }  else {ebi = ebi.Erase(); modified=true;}
>
>
Done.


>  ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137831813>
> :
>
> > +    worklist.pop();
> +    block->ForEachSuccessorLabel(
> +        [&reachable_blocks, &worklist, this](uint32_t label_id) {
> +          auto successor = label2block_[label_id];
> +          reachable_blocks.insert(successor);
> +          worklist.push(successor);
> +        });
> +  }
> +
> +  // Erase unreachable blocks.
> +  for (auto ebi = func->begin(); ebi != func->end();)
> +    if (reachable_blocks.find(&*ebi) == reachable_blocks.end()) {
> +      ebi = ebi.Erase();
> +      modified = true;
> +    } else {
> +      ebi++;
>
> Always prefer pre-order increment/decrement on non-primitive types.
> Iterators in particular. This saves a copy.
>
It doesn't, given the implementation of the iterator classes, but changed
anyway (and nowadays, I do expect a good optimizing compiler to figure this
one out on its own, but it's not always the case ;).


> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137843639>
> :
>
> > +  worklist.push(func->entry().get());
> +  while (!worklist.empty()) {
> +    ir::BasicBlock* block = worklist.front();
> +    worklist.pop();
> +    block->ForEachSuccessorLabel(
> +        [&reachable_blocks, &worklist, this](uint32_t label_id) {
> +          auto successor = label2block_[label_id];
> +          reachable_blocks.insert(successor);
> +          worklist.push(successor);
> +        });
> +  }
> +
> +  // Erase unreachable blocks.
> +  for (auto ebi = func->begin(); ebi != func->end();)
> +    if (reachable_blocks.find(&*ebi) == reachable_blocks.end()) {
> +      ebi = ebi.Erase();
>
> You're too used to LLVM. Remember that in the current ir::* class
> hierarchy, instructions only have IDs for the things they use, and the
> DefUseManager is a separate layer over it.
>
Gah!  Yes.  So, this is something that would be  awesome if we can change.
Some ideas coming up in the next few days.

So deleting an Instruction will not delete the references to that
> instruction.
> Thinks you'll have to clean up:
>
>    - OpName, possibly OpMemberName (for struct types, so does not apply
>    in this pass)
>    - OpDecoration, OpDecorationGroup
>    - OpPhi, e.g. you delete A, and so all values defined in A, but that
>    block and its values might be used in an OpPhi in a live block.
>
> Use DefUseManager to trace uses. It has a convenience KillInst that
> converts an existing instruction to a Nop and kill the uses. That'a a
> common idiom you'll see: KillInst, then the Module's write-to-output
> function can be used to skip over all nops. It works. However, cleaning up
> blocks is another matter. Check Greg's dead-branch-elim code for examples
> of how to clean up blocks.
>
Thanks.  Done.


> And then we have to agree on whether this pass promises to keep up the
> DefUseManager up to date. By default it does not have to make that promise.
> But our direction is to "save time" by agreeing to keep it up to date, and
> then have the pass manager track whether the DefUseManager is stale or not.
>
OK.   This is part of another conversation, but I think we can make some
improvements here that will mean not having to update all these satellite
data structures when doing basic operations like removing instructions,
edges or basic blocks.

------------------------------
>
> In include/spirv-tools/optimizer.hpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137844149>
> :
>
> > @@ -393,6 +393,13 @@ Optimizer::PassToken CreateAggressiveDCEPass();
>  // The pass remaps result ids to a compact and gapless range starting from %1.
>  Optimizer::PassToken CreateCompactIdsPass();
>
> +// Creates a CFG cleanup pass.
> +// This pass removes cruft from the control flow graph. It currently includes
> +// the following functionality:
>
> The pass only operates on functions reachable from some entry point.
> Should say so.
>

Done.


> ------------------------------
>
> In test/opt/cfg_cleanup_test.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137844637>
> :
>
> > +OpSelectionMerge %15 None
> +OpBranch %16
> +%16 = OpLabel
> +%20 = OpLoad %float %inf
> +%21 = OpFAdd %float %20 %float_n0_5
> +%22 = OpCompositeConstruct %v4float %21 %21 %21 %21
> +OpStore %outf4 %22
> +OpBranch %15
> +%15 = OpLabel
> +OpReturn
> +OpFunctionEnd
> +)";
> +
> +  SinglePassRunAndCheck<opt::CFGCleanupPass>(
> +      declarations + body_before, declarations + body_after, true, true);
> +}
>
> Need cases for:
>
>    - showing cleanup of OpName and OpDecoration, OpGroupDecorate
>
>
Added a test for this one, but I'm causing some failures during assembly
that I'm still trying to figure out.


>
>    - showing cleanup of OpPhis in live blocks that remain
>
> Well, I'm not sure how I can get this with this change.  This cleanup does
not flatten out branches nor multi-branches. Perhaps from non-structured
control flow we could get this.


>
>    - keeping unreachable merge block
>    - keeping unreachable continue target, and the loop backedge it
>    dominates even if it's not in the same block
>
>  Will do.


>    - in these latter two cases: non-trivial defs and uses inside those
>    unreachable blocks
>
> Hmm, ok, but when a block is gone, we remove all its instructions.  This
is covered by  the existing tests?  Or are you thinking something else?


>
> ------------------------------
>
> In tools/opt/opt.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r137844700>
> :
>
> > @@ -112,9 +112,12 @@ NOTE: The optimizer is a work in progress.
>                 Join two blocks into a single block if the second has the
>                 first as its only predecessor. Performed only on entry point
>                 call tree functions.
> -  -h, --help
> +  --cfg-cleanup
> +               Cleanup the control flow graph. This will remove any unnecessary
> +               code from the CFG like unreachable code.
>
> .... "Peformed only on entry point call tree functions"
>

Done.

dnovillo commented:
On Thu, Sep 21, 2017 at 4:50 PM, Diego Novillo <dnovillo@google.com> wrote:

>
> Added a test for this one, but I'm causing some failures during assembly
> that I'm still trying to figure out.
>

Done.  I had written the wrong assembly.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,117 @@
+  }
+
+  // Erase unreachable blocks.
+  for (auto ebi = func->begin(); ebi != func->end();)
```
Strongly prefer braces around multi-line bodies.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,117 @@
+    KillNamesAndDecorates(inst);
+    def_use_mgr_->KillInst(inst);
+  });
+  *bi = bi->Erase();
```
Need to clean up phis  in live blocks that have this block as a predecessor.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,117 @@
+
+  // Process all entry point functions.
+  ProcessFunction pfn = [this](ir::Function* fp) { return CFGCleanup(fp); };
+  bool modified = ProcessEntryPointCallTree(pfn, module);
```
I suggest generalizing this pass to use ProcessReachableCallTree that @s-perron recently added.
That is, also cover functions reachable from exported functions.

Would also have to update the pass description.

dnovillo replied:
Oops. Forgot this.  I'm adding this now.

dnovillo replied:
OK, now done.

dnovillo commented:
On Tue, Sep 26, 2017 at 5:28 PM, David Neto <notifications@github.com>
wrote:
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r141124881>
> :
>
> > +  };
> +
> +  // Transitively mark all blocks reachable from the entry as reachable.
> +  while (!worklist.empty()) {
> +    ir::BasicBlock* block = worklist.front();
> +    worklist.pop();
> +
> +    // All the successors of a live block are also live.
> +    block->ForEachSuccessorLabel(mark_reachable);
> +
> +    // All the Merge and ContinueTarget blocks of a live block are also live.
> +    block->ForMergeAndContinueLabel(mark_reachable);
> +  }
> +
> +  // Erase unreachable blocks.
> +  for (auto ebi = func->begin(); ebi != func->end();)
>
> Strongly prefer braces around multi-line bodies.
>

Done.


> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r141126105>
> :
>
> > +
> +#include "cfg_cleanup_pass.h"
> +
> +#include "function.h"
> +#include "module.h"
> +
> +namespace spvtools {
> +namespace opt {
> +
> +void CFGCleanupPass::RemoveBlock(ir::Function::iterator* bi) {
> +  auto &block = **bi;
> +  block.ForEachInst([this](ir::Instruction* inst) {
> +    KillNamesAndDecorates(inst);
> +    def_use_mgr_->KillInst(inst);
> +  });
> +  *bi = bi->Erase();
>
> Need to clean up phis in live blocks that have this block as a predecessor.
>

Done.  PTAL.

This reminds me that we'll want to add more CFG cleanup including removal
of degenerate Phis.  These are particularly common after propagation and
dce.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,170 @@
+  });
+
+  // Remove the label instruction last.
+  def_use_mgr_->KillInst(block.GetLabelInst());
```
What if the label has a name?

This code will produce invalid SPIR-V when run on the following input:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main"
               OpSource GLSL 430
               OpName %main "main"
               OpName %dead "dead"
       %void = OpTypeVoid
          %5 = OpTypeFunction %void
       %main = OpFunction %void None %5
          %6 = OpLabel
               OpReturn
       %dead = OpLabel
               OpReturn
               OpFunctionEnd
```

%dead is dead, and will be swept away.  But then the OpName targeting %dead will be left around.

Should use KillNamesAndDecorates on the label instruction.

Please add the above as a test case too.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,170 @@
+
+    ir::Instruction* phi_inst = u.inst;
+    std::vector<ir::Operand> keep_operands;
+    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
```
Start i from 2 here, then you won't need the test against i >= 2 later on.
(But also move the "Phi arguments start at index 2" comment to just before the loop)

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,170 @@
+        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
+        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
+               "Phi operands should have exactly one word in them.");
+        uint32_t var_id = var_op.words.front();
```
It's a little more idiomatic t write:

```
 uint32_t var_id = GetSingleWordOperand(i);
 uint32_t block_id = GetSingleWordOperand(i+1);
```
rather than accessing the guts of an Operand instance.

Separately, both var_id and block_id should be const.

dnovillo commented:
I think this is ready for review again.  Any other changes?  Thanks.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,170 @@
+        }
+      }
+
+      keep_operands.push_back(var_op);
```
Where's block_op ?  This smells like a bug.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,170 @@
+        uint32_t var_id = var_op.words.front();
+        uint32_t block_id = block_op.words.front();
+        if (var_id == inst_id && block_id == block.id()) {
+          i++;
```
This increment looks wrong.

I think you'll get out of phase. 

I recommend:
- the "for" statement should be:    `for (i = 2 ; i < phi_inst->NumOperands() ; i+= 2)`
- then the body of the loop either adds the var_op and block_op as a pair, or not.

dnovillo commented:
On Thu, Oct 5, 2017 at 2:55 PM, David Neto <notifications@github.com> wrote:

> What if the label has a name?
>
> This code will produce invalid SPIR-V when run on the following input:
>
>                OpCapability Shader
>           %1 = OpExtInstImport "GLSL.std.450"
>                OpMemoryModel Logical GLSL450
>                OpEntryPoint Vertex %main "main"
>                OpSource GLSL 430
>                OpName %main "main"
>                OpName %dead "dead"
>        %void = OpTypeVoid
>           %5 = OpTypeFunction %void
>        %main = OpFunction %void None %5
>           %6 = OpLabel
>                OpReturn
>        %dead = OpLabel
>                OpReturn
>                OpFunctionEnd
>
> %dead is dead, and will be swept away. But then the OpName targeting %dead
> will be left around.
>
> Should use KillNamesAndDecorates on the label instruction.
>
> Please add the above as a test case too.
>

Done.


> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r142958693>
> :
>
> > +    return;
> +  }
> +
> +  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
> +  if (uses == nullptr) {
> +    return;
> +  }
> +
> +  for (auto u : *uses) {
> +    if (u.inst->opcode() != SpvOpPhi) {
> +      continue;
> +    }
> +
> +    ir::Instruction* phi_inst = u.inst;
> +    std::vector<ir::Operand> keep_operands;
> +    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
>
> Start i from 2 here, then you won't need the test against i >= 2 later on.
> (But also move the "Phi arguments start at index 2" comment to just before
> the loop)
>

Can't.  We need to reproduce the full operand vector for the Phi.  Only the
unwanted operands need to be removed. The test for i>=2 is needed so we
only match the actual Phi operands.  This makes me think we may want to
have sub-classes of ir::Instruction to avoid this kind of thing.  It's
slightly odd that each Phi logical operand is actually 2 operands.


> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r142959318>
> :
>
> > +  for (auto u : *uses) {
> +    if (u.inst->opcode() != SpvOpPhi) {
> +      continue;
> +    }
> +
> +    ir::Instruction* phi_inst = u.inst;
> +    std::vector<ir::Operand> keep_operands;
> +    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
> +      const ir::Operand& var_op = phi_inst->GetOperand(i);
> +      if (i >= 2 && i < phi_inst->NumOperands() - 1) {
> +        // PHI arguments start at index 2. Each argument consists of two
> +        // operands: the variable id and the originating block id.
> +        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
> +        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
> +               "Phi operands should have exactly one word in them.");
> +        uint32_t var_id = var_op.words.front();
>
> It's a little more idiomatic t write:
>
>  uint32_t var_id = GetSingleWordOperand(i);
>  uint32_t block_id = GetSingleWordOperand(i+1);
>
> rather than accessing the guts of an Operand instance.
>
> Separately, both var_id and block_id should be const.
>
 Done.

------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r142961769>
> :
>
> > +      const ir::Operand& var_op = phi_inst->GetOperand(i);
> +      if (i >= 2 && i < phi_inst->NumOperands() - 1) {
> +        // PHI arguments start at index 2. Each argument consists of two
> +        // operands: the variable id and the originating block id.
> +        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
> +        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
> +               "Phi operands should have exactly one word in them.");
> +        uint32_t var_id = var_op.words.front();
> +        uint32_t block_id = block_op.words.front();
> +        if (var_id == inst_id && block_id == block.id()) {
> +          i++;
> +          continue;
> +        }
> +      }
> +
> +      keep_operands.push_back(var_op);
>
> Where's block_op ? This smells like a bug.
>

Hmm?  It's the operand right after the variable.  I think we need a better
API to Phi operands.  I'll try to re-work this loop to be less confusing.


> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r142961903>
> :
>
> > +    }
> +
> +    ir::Instruction* phi_inst = u.inst;
> +    std::vector<ir::Operand> keep_operands;
> +    for (uint32_t i = 0; i < phi_inst->NumOperands(); i++) {
> +      const ir::Operand& var_op = phi_inst->GetOperand(i);
> +      if (i >= 2 && i < phi_inst->NumOperands() - 1) {
> +        // PHI arguments start at index 2. Each argument consists of two
> +        // operands: the variable id and the originating block id.
> +        const ir::Operand& block_op = phi_inst->GetOperand(i + 1);
> +        assert(var_op.words.size() == 1 && block_op.words.size() == 1 &&
> +               "Phi operands should have exactly one word in them.");
> +        uint32_t var_id = var_op.words.front();
> +        uint32_t block_id = block_op.words.front();
> +        if (var_id == inst_id && block_id == block.id()) {
> +          i++;
>
> This increment looks wrong.
>
> I think you'll get out of phase.
>
It doesn't, but I agree that the code is confusing.  I'll re-work it to not
be so opaque.

> I recommend:
>
>    - the "for" statement should be: for (i = 2 ; i <
>    phi_inst->NumOperands() ; i+= 2)
>
> Can't do this.  All operands need to be pushed into the new operand
vector.

>
>    - then the body of the loop either adds the var_op and block_op as a
>    pair, or not.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#pullrequestreview-67396233>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG9RZ5LJF183YpBaeJglbBCouthQBVMQks5spO3hgaJpZM4PQDKj>
> .
>

dneto0 commented:
> Can't.  We need to reproduce the full operand vector for the Phi. 

I see.   I get it now.   I think it would be clearer to move the i++ out of the for statement and only do it in the cases inside the loop.  So you'd have something like:

```
 if ( i < 2 ) {
  keep_operands.push_back( ...);
  ++i;
 } else {
    // handle (var, pred) case, as a pair
    either keep both or neither
    i += 2;
}
```

> >  Where's block_op ? This smells like a bug.

> Hmm?  It's the operand right after the variable.  I think we need a better
API to Phi operands.  I'll try to re-work this loop to be less confusing.

Right.  My problem here is that the code at the end of the function replaces all the (var, predecessor) operands with the sequence of operands contained in keep_operands.   So you have to put both the var_op and the block_op.  I think you were missing the block_op?

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  // Traverse all the operands in |phi|. Build the new operand vector by adding
+  // all the original operands from |phi| except the unwanted one.
+  for (uint32_t i = 0; i < phi->NumOperands();) {
```
Thanks.  This is much easier to understand.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,186 @@
+
+  for (auto u : *uses) {
+    if (u.inst->opcode() == SpvOpPhi) {
+      RemovePhiOperand(u.inst, inst_id, block.id());
```
In SPIR-V the block in a Phi is the immediate predecessor block, not necessarily the block in which the associated value is defined.

Example:

%block0  defines value %e
Then %block0 branches into %block1, then %block1 branches to %block2. 
%block2 has a phi with (%e, %block1).
Then when killing %block0 you're going to delete the definition of %e but miss removing (%e, %block1) from the phi.

So you need an alternative case where you substitute an undef value for inst used in a subsequent phi.
E.g. something like in https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/opt/local_ssa_elim_pass.cpp#L271 where an undef value is generated for the given type.

You'll need a test case for this scenario.

dneto0 commented:
Also, please rebase against master.  :-)

dnovillo commented:
On Tue, Oct 10, 2017 at 12:37 PM, David Neto <notifications@github.com>
wrote:

> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r143782193>
> :
>
> > +
> +void CFGCleanupPass::RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
> +                                                  ir::Instruction* inst) {
> +  uint32_t inst_id = inst->result_id();
> +  if (inst_id == 0) {
> +    return;
> +  }
> +
> +  analysis::UseList* uses = def_use_mgr_->GetUses(inst_id);
> +  if (uses == nullptr) {
> +    return;
> +  }
> +
> +  for (auto u : *uses) {
> +    if (u.inst->opcode() == SpvOpPhi) {
> +      RemovePhiOperand(u.inst, inst_id, block.id());
>
> In SPIR-V the block in a Phi is the immediate predecessor block, not
> necessarily the block in which the associated value is defined.
>
> Example:
>
> %block0 defines value %e
> Then %block0 branches into %block1, then %block1 branches to %block2.
> %block2 has a phi with (%e, %block1).
> Then when killing %block0 you're going to delete the definition of %e but
> miss removing (%e, %block1) from the phi.
>
> So you need an alternative case where you substitute an undef value for
> inst used in a subsequent phi.
> E.g. something like in https://github.com/KhronosGrou
> p/SPIRV-Tools/blob/master/source/opt/local_ssa_elim_pass.cpp#L271 where
> an undef value is generated for the given type.
>
> You'll need a test case for this scenario.
>

I'm thinking... What if I just match the value?  It's an SSA name, so I
just need to match it.  If the same value is referenced in multiple
arguments, then they all need to be removed.  Pseudocode (apologies)
%b6: jump %b5
%b0: %20 = ...
%b5: switch (%3) {
  %b1: case 0:
     [ ... ] (no new name for %20)
     break;
  %b2: case 10:
     [...] (no new name for %20)
   case 23:
     %b3: %21 = ...
}
%22 = phi(%20 %b0, %20 %b1, %20 %b2, %21 %b3)

This code should remove all the phi arguments for %20, right?  Just
matching %20 should be enough?

dneto0 commented:
> %22 = phi(%20 %b0, %20 %b1, %20 %b2, %21 %b3)

> This code should remove all the phi arguments for %20, right?  Just
matching %20 should be enough?

Yes, but there are two cases: 
-  The phi parent block is the block you are removing.  Remove the pair (%20 %b0)
- The phi parent block is not the block you are removing.  Replace the %20 with an Undef of the right type.  (%u %b1) etc. where %u is defined appropriately.

dneto0 commented:
Stop the presses.  I thought about this further, and it's kind of weird.

Let's look at that second case again:  The phi parent block is not the block you are removing. 
From your example,  %20 %b1 is one such pair.   Now, if value %20 reaches %22 via %b1, then a use of %20 could exist in %b1. That means the definition of %20 dominates %b1.  But you're removing %20 because it's in a dead block %b0.     That implies that %b0 must be dead too.  So you shouldn't have to bother replacing %20 by an undef in the pair %20 %b1 because you're going to eventually remove the pair anyway.

dneto0 commented on test/opt/cfg_cleanup_test.cpp:
```diff
@@ -0,0 +1,272 @@
+         %24 = OpISub %int %11 %int_23
+               OpBranch %17
+         %17 = OpLabel
+         %31 = OpPhi %int %20 %16 %24 %21
```
This case proves that the pair (%20 %16) is removed.  But it's removed because the value %20 is defined in a removed block.

You should also have a case where you remove a pair because the block is dead.  E.g. if you had

constant:  %int_0  = OpConstant %int 0

and then you had 
 %31 = OpPhi %int %int_0 %16 %24 %21

This should also result in removing the pair (%int_0 %16)

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,186 @@
+  // Remove the label instruction last.
+  auto label = block.GetLabelInst();
+  KillNamesAndDecorates(label);
+  def_use_mgr_->KillInst(label);
```
You'll also need to remove this label from any phis that might use it.
E.g. if removing block A, then remove any pairs (X A) from any Phis.   For example X could be a constant value (as in the example I gave earlier.)

dneto0 replied:
Still need to take care of this.

If RemovePhiOperand allowed a 0 arg id then you could use that.

dnovillo commented:
So, I think I've convinced myself that we cannot remove a PHI argument
unless both the block (i.e., the label) is being removed.  In the example I
gave:

%b6: jump %b5
%b0: %20 = ...
%b5: switch (%3) {
  %b1: case 0:
     [ ... ] (no new name for %20)
     break;
  %b2: case 10:
     [...] (no new name for %20)
  %b3: case 23:
     %21 = ...
}
%22 = phi(%20 %b0, %20 %b1, %20 %b2, %21 %b3)

In here, we are removing %b0, so %20 is being removed.  Removing %b0 means
that the PHI node %22 gets its in-degree reduced by 1.  So instead of 4
pairs of (value, block) arguments, it should have 3.

%22 = phi(%20 %b0, %20 %b1,    %20 %b2,    %21 %b3) must be converted to
%22 = phi(         %Undef %b1, %Undef %b2, %21 %b3).

We cannot remove the arguments for %b1 nor %b2, because those blocks are
still there.  We'd get an invalid PHI, if we did.  So, I think the right
way to update PHI nodes is what you had proposed originally:


   1. If the argument's block matches an unreachable node, the argument
   must be removed.
   2. If the argument's value matches a removed value, but the block is
   still live, then the argument's value must be replaced with Undef.


Does this sound reasonable?

On Wed, Oct 11, 2017 at 9:45 AM, David Neto <notifications@github.com>
wrote:

> *@dneto0* commented on this pull request.
> ------------------------------
>
> In test/opt/cfg_cleanup_test.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r144015407>
> :
>
> > +     %int_23 = OpConstant %int 23
> +      %int_5 = OpConstant %int 5
> +%_ptr_Output_int = OpTypePointer Output %int
> +    %outparm = OpVariable %_ptr_Output_int Output
> +       %main = OpFunction %void None %3
> +          %5 = OpLabel
> +         %11 = OpLoad %int %y
> +               OpBranch %21
> +         %16 = OpLabel
> +         %20 = OpIAdd %int %11 %int_42
> +               OpBranch %17
> +         %21 = OpLabel
> +         %24 = OpISub %int %11 %int_23
> +               OpBranch %17
> +         %17 = OpLabel
> +         %31 = OpPhi %int %20 %16 %24 %21
>
> This case proves that the pair (%20 %16) is removed. But it's removed
> because the value %20 is defined in a removed block.
>
> You should also have a case where you remove a pair because the block is
> dead. E.g. if you had
>
> constant: %int_0 = OpConstant %int 0
>
> and then you had
> %31 = OpPhi %int %int_0 %16 %24 %21
>
> This should also result in removing the pair (%int_0 %16)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#pullrequestreview-68610907>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG9RZ1GUeFHxLcDxJhMFpRzKUHzhwfG_ks5srMZjgaJpZM4PQDKj>
> .
>

dnovillo commented:
I implemented this approach.  PTAL.

Thanks.

On Wed, Oct 11, 2017 at 2:35 PM, Diego Novillo <dnovillo@google.com> wrote:

> So, I think I've convinced myself that we cannot remove a PHI argument
> unless both the block (i.e., the label) is being removed.  In the example I
> gave:
>
> %b6: jump %b5
> %b0: %20 = ...
> %b5: switch (%3) {
>   %b1: case 0:
>      [ ... ] (no new name for %20)
>      break;
>   %b2: case 10:
>      [...] (no new name for %20)
>   %b3: case 23:
>      %21 = ...
> }
> %22 = phi(%20 %b0, %20 %b1, %20 %b2, %21 %b3)
>
> In here, we are removing %b0, so %20 is being removed.  Removing %b0 means
> that the PHI node %22 gets its in-degree reduced by 1.  So instead of 4
> pairs of (value, block) arguments, it should have 3.
>
> %22 = phi(%20 %b0, %20 %b1,    %20 %b2,    %21 %b3) must be converted to
> %22 = phi(         %Undef %b1, %Undef %b2, %21 %b3).
>
> We cannot remove the arguments for %b1 nor %b2, because those blocks are
> still there.  We'd get an invalid PHI, if we did.  So, I think the right
> way to update PHI nodes is what you had proposed originally:
>
>
>    1. If the argument's block matches an unreachable node, the argument
>    must be removed.
>    2. If the argument's value matches a removed value, but the block is
>    still live, then the argument's value must be replaced with Undef.
>
>
> Does this sound reasonable?
>
> On Wed, Oct 11, 2017 at 9:45 AM, David Neto <notifications@github.com>
> wrote:
>
>> *@dneto0* commented on this pull request.
>> ------------------------------
>>
>> In test/opt/cfg_cleanup_test.cpp
>> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r144015407>
>> :
>>
>> > +     %int_23 = OpConstant %int 23
>> +      %int_5 = OpConstant %int 5
>> +%_ptr_Output_int = OpTypePointer Output %int
>> +    %outparm = OpVariable %_ptr_Output_int Output
>> +       %main = OpFunction %void None %3
>> +          %5 = OpLabel
>> +         %11 = OpLoad %int %y
>> +               OpBranch %21
>> +         %16 = OpLabel
>> +         %20 = OpIAdd %int %11 %int_42
>> +               OpBranch %17
>> +         %21 = OpLabel
>> +         %24 = OpISub %int %11 %int_23
>> +               OpBranch %17
>> +         %17 = OpLabel
>> +         %31 = OpPhi %int %20 %16 %24 %21
>>
>> This case proves that the pair (%20 %16) is removed. But it's removed
>> because the value %20 is defined in a removed block.
>>
>> You should also have a case where you remove a pair because the block is
>> dead. E.g. if you had
>>
>> constant: %int_0 = OpConstant %int 0
>>
>> and then you had
>> %31 = OpPhi %int %int_0 %16 %24 %21
>>
>> This should also result in removing the pair (%int_0 %16)
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#pullrequestreview-68610907>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AG9RZ1GUeFHxLcDxJhMFpRzKUHzhwfG_ks5srMZjgaJpZM4PQDKj>
>> .
>>
>
>

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,269 @@
+//    the argument value with OpUndef.  Note that there may be several
+//    operands in |phi| that meet this requirement.
+//
+// For example, in the switch() below, assume that we want to remove the
```
Good example!

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,269 @@
+  return undefId;
+}
+
+// Remove Phi operand (|rm_arg_id|, |rm_block_id|) from |phi|. There are two
```
Excellent comment.

dneto0 commented on source/opt/cfg_cleanup_pass.h:
```diff
@@ -0,0 +1,88 @@
+  void RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
+                                    ir::Instruction* inst);
+
+  // Remove Phi operand (|rm_arg_id|, |rm_block_id|) from |phi|.
```
This needs slightly more explanation?
```
Remove Phi operand involving either |rm_arg_id| or |rm_block_id|:
- Remove any pair (var, |rm_block_id|)
- Otherwise, replace (|rm_arg_id|, block) with ( U, block ) where U is the undefined value with same type as |rm_arg_id|
```

dneto0 replied:
It would also be handy if you allowed |rm_arg_id| to be zero.  Then you can use it for the cleanup of OpLabel mentioned in phis, e.g. where the value is some OpConstant.

dneto0 commented:
The Travis build failure looks to be unrelated.  I'm investigating that.

dnovillo commented:
On Fri, Oct 13, 2017 at 4:52 PM, David Neto <notifications@github.com>
wrote:

> In source/opt/cfg_cleanup_pass.h
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r144603538>
> :
>
> > +  // Remove all the unreachable basic blocks in |func|.
> +  bool RemoveUnreachableBlocks(ir::Function* func);
> +
> +  // Remove the block pointed by the iterator |*bi|. This also removes
> +  // all the instructions in the pointed-to block.
> +  void RemoveBlock(ir::Function::iterator *bi);
> +
> +  // Initialize the pass.
> +  void Initialize(ir::Module* module);
> +
> +  // Remove the result from |inst| from every Phi instruction reached by
> +  // |inst|. The instruction is coming from basic block |block|.
> +  void RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
> +                                    ir::Instruction* inst);
> +
> +  // Remove Phi operand (|rm_arg_id|, |rm_block_id|) from |phi|.
>
> This needs slightly more explanation?
>
> Remove Phi operand involving either |rm_arg_id| or |rm_block_id|:
> - Remove any pair (var, |rm_block_id|)
> - Otherwise, replace (|rm_arg_id|, block) with ( U, block ) where U is the undefined value with same type as |rm_arg_id|
>
>
I've re-implemented this code.  It turns out that we need to analyze *ALL*
the Phi instructions in the program.  Before, the code was trying to be
selective and only analyze Phis reached by unreachable blocks, but that's
not always the case. A Phi operand may be reached by an unreachable block
that has no def-use edges into it.  I added a test for it:

       %main = OpFunction %void None %3
          %5 = OpLabel
               OpBranch %14
         %40 = OpLabel
          %9 = OpLoad %int %y
         %12 = OpSGreaterThan %bool %9 %int_10
               OpSelectionMerge %14 None
               OpBranchConditional %12 %13 %14
         %13 = OpLabel
               OpBranch %14
         %14 = OpLabel
         %25 = OpPhi %int %24 %5 %int_23 %13

Block %13 here is unreachable, but the argument %int_23 is not defined
there.  So, by just following def-use edges, we would never analyze that
Phi instruction.

 ------------------------------
>
> In source/opt/cfg_cleanup_pass.h
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r144604386>
> :
>
> > +  // Remove all the unreachable basic blocks in |func|.
> +  bool RemoveUnreachableBlocks(ir::Function* func);
> +
> +  // Remove the block pointed by the iterator |*bi|. This also removes
> +  // all the instructions in the pointed-to block.
> +  void RemoveBlock(ir::Function::iterator *bi);
> +
> +  // Initialize the pass.
> +  void Initialize(ir::Module* module);
> +
> +  // Remove the result from |inst| from every Phi instruction reached by
> +  // |inst|. The instruction is coming from basic block |block|.
> +  void RemoveFromReachedPhiOperands(const ir::BasicBlock& block,
> +                                    ir::Instruction* inst);
> +
> +  // Remove Phi operand (|rm_arg_id|, |rm_block_id|) from |phi|.
>
> It would also be handy if you allowed |rm_arg_id| to be zero. Then you can
> use it for the cleanup of OpLabel mentioned in phis, e.g. where the value
> is some OpConstant.
>

The new code should handle this case.  The new test is an example of this.



> ------------------------------
>
> In source/opt/cfg_cleanup_pass.cpp
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#discussion_r144604756>
> :
>
> > +  auto& block = **bi;
> +  block.ForEachInst([&block, this](ir::Instruction* inst) {
> +    // Note that we do not kill the block label instruction here. The label
> +    // instruction is needed to identify the block, which is needed by the
> +    // removal of PHI operands.
> +    if (inst != block.GetLabelInst()) {
> +      RemoveFromReachedPhiOperands(block, inst);
> +      KillNamesAndDecorates(inst);
> +      def_use_mgr_->KillInst(inst);
> +    }
> +  });
> +
> +  // Remove the label instruction last.
> +  auto label = block.GetLabelInst();
> +  KillNamesAndDecorates(label);
> +  def_use_mgr_->KillInst(label);
>
> Still need to take care of this.
>
> If RemovePhiOperand allowed a 0 arg id then you could use that.
>

Sorry.  Take care of what?  I'm not following you here.

PTAL.  Thanks.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,287 @@
+
+  // Traverse all the operands in |phi|. Build the new operand vector by adding
+  // all the original operands from |phi| except the unwanted ones.
+  bool undef_created = false;
```
You could use the convention that undef_id > 0 if and only if you've created the undef value.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,287 @@
+
+  // Mark reachable all blocks reachable from the function's entry block.
+  std::unordered_set<ir::BasicBlock*> reachable_blocks;
+  std::unordered_set<ir::BasicBlock*> visited_blocks;
```
It seems to me that you could remove visited_blocks in favour of using reachable_blocks.  But it's not that important, and this way is readable.

dneto0 commented on source/opt/cfg_cleanup_pass.cpp:
```diff
@@ -0,0 +1,287 @@
+
+    // If the block is reachable and has Phi instructions, remove all
+    // operands from its Phi instructions that reference unreachable blocks.
+    if (block.HasPhiInstructions()) {
```
It's redundant to check HasPhiInstructions  here.  Just run ForEachPhiInst directly, and be happy if it never calls the lambda.

dneto0 commented:
I forgot.  This should also add source/opt/cfg_cleanup_pass.cpp to the Android.mk.  I'll do that before merging.

dneto0 commented:
hm. I think I merged the wrong thing. :-(

dnovillo commented:
Anything for me to fix?

On Thu, Oct 19, 2017 at 7:19 PM, David Neto <notifications@github.com>
wrote:

> hm. I think I merged the wrong thing. :-(
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#issuecomment-338009843>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG9RZ3M0rcy1TpW1cQen3Mta18-3kNj2ks5st6CygaJpZM4PQDKj>
> .
>

dneto0 commented:
Ok. I pushed in the right thing now.  :-)

Rebased and merged into master as  c75704ec and fixup 8ec62deb2

dneto0 commented:
I think what happened is that I had a stale 702 branch lying around from a previous iteration of this review.  I've now put on the actual head of your 702, suitably rebased.  So all good now.  First time this has happened to me.  :-(

dnovillo commented:
Cool.  Thanks!

On Thu, Oct 19, 2017 at 7:33 PM, David Neto <notifications@github.com>
wrote:

> I think what happened is that I had a stale 702 branch lying around from a
> previous iteration of this review. I've now put on the actual head of your
> 702, suitably rebased. So all good now. First time this has happened to me.
> :-(
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/KhronosGroup/SPIRV-Tools/pull/810#issuecomment-338013390>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AG9RZ-hJoGoUkwRMZZlBt_LNb34Or3hZks5st6QcgaJpZM4PQDKj>
> .
>
